### ISSUE TITLE
Pushwoosh and Firebase Messaging

### ISSUE THREAD

**Issue by Andrekarma** [Pushwoosh/pushwoosh-flutter#70]
i have flutter application, and i want to receive push notifications from Pushwoosh but also from Firebase.
If i set both the sdk i can only receive the push notifications from the Firebase platform
Do you have a solution?

**Comment by wfhm** [comment]
@Andrekarma Do you want to receive pushes from Pushwoosh and Firebase on iOS, Android, or on both platforms?

To do that, you have to do some native coding:
* For Android you have to add custom Notification Service Extension that will route notifications between Firebase and Pushwoosh. You can find a guide here:
https://docs.pushwoosh.com/platform-docs/pushwoosh-sdk/android-push-notifications/android-faq#using-pushwoosh-sdk-with-other-fcm-services

You will need to call methods of Firebase plugin in your custom service (specifically, methods of [FlutterFirebaseMessagingReceiver](https://github.com/firebase/flutterfire/blob/master/packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/FlutterFirebaseMessagingReceiver.java) and [FlutterFirebaseMessagingService](https://github.com/firebase/flutterfire/blob/master/packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/FlutterFirebaseMessagingService.java))

* Both Pushwoosh and Firebase plugins swizzle delegate methods in iOS, resulting in only one delegate working properly (pushes from both Firebase and Pushwoosh will be received but callback will work only for one of the plugins). To work around this, you can disable Pushwoosh delegate and modify AppDelegate in your project as below:

1. Comment out or remove the following lines in PushwooshPlugin.m file:
```
if (@available(iOS 10, *)) {
    [UNUserNotificationCenter currentNotificationCenter].delegate = [PushNotificationManager pushManager].notificationCenterDelegate;
}
```
You can open this file in your Xcode project here:
<img width="595" alt="Screenshot 2023-10-03 at 20 57 50" src="https://github.com/Pushwoosh/pushwoosh-flutter/assets/13344781/8eb22d32-be06-4b7c-8e01-da6d8e399622">
2. Modify your AppDelegate as below:
```
import UIKit
import Flutter
import FirebaseMessaging
import Pushwoosh

@UIApplicationMain
@objc class AppDelegate: FlutterAppDelegate {
  override func application(
    _ application: UIApplication,
    didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
  ) -> Bool {
    GeneratedPluginRegistrant.register(with: self)
      if #available(iOS 10.0, *) {
          UNUserNotificationCenter.current().delegate = self as UNUserNotificationCenterDelegate
      }
    return super.application(application, didFinishLaunchingWithOptions: launchOptions)
  }
    
    override func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
        Messaging.messaging().apnsToken = deviceToken;
        Pushwoosh.sharedInstance().handlePushRegistration(deviceToken)
        super.application(application, didRegisterForRemoteNotificationsWithDeviceToken: deviceToken)
      }
        
      override func userNotificationCenter(_ center: UNUserNotificationCenter, willPresent notification: UNNotification, withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
        let userInfo = notification.request.content.userInfo
          print("WILL PRESENT NOTIFICATION: " + notification.request.content.body)

          Pushwoosh.sharedInstance().notificationCenterDelegateProxy?.defaultNotificationCenterDelegate.userNotificationCenter?(center, willPresent: notification, withCompletionHandler: completionHandler)

        // Uncomment this line to receive notifications on foreground
        // completionHandler([[.alert, .sound]])
        super.userNotificationCenter(center, willPresent: notification, withCompletionHandler: completionHandler)
      }
        
      override func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
        let userInfo = response.notification.request.content.userInfo
       
        Pushwoosh.sharedInstance().notificationCenterDelegateProxy?.defaultNotificationCenterDelegate.userNotificationCenter?(center, didReceive: response, withCompletionHandler: completionHandler)
          print("DID RECEIVE NOTIFICATION: " + response.notification.request.content.body)
        completionHandler()
        super.userNotificationCenter(center, didReceive: response, withCompletionHandler: completionHandler)
      }
        
      override func application(_ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable : Any],
       fetchCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void) {
        completionHandler(.noData)
        super.application(application, didReceiveRemoteNotification: userInfo, fetchCompletionHandler: completionHandler)
      }
}
```

Please let us know about the results

**Comment by enginseer-dev** [comment]
Hi @Andrekarma, I hope the solution we provided earlier has been helpful. Just to give you another option, we've recently released an updated version of our SDK - version 2.2.16. This update might simplify things for you, allowing for seamless integration of push notifications from both Pushwoosh and Firebase in your application.

You can check out the latest SDK here: [Pushwoosh Flutter SDK 2.2.16](https://github.com/Pushwoosh/pushwoosh-flutter/releases/tag/2.2.16)

Please let us know if this resolves your issue or if you need further assistance.
